frontend/src/main/java/com/google/tripmeout/frontend/servlets/PlaceParentServlet.java

```diff
@@ -0,0 +1,110 @@
+package com.google.tripmeout.frontend.servlets;
+
+import com.google.common.flogger.FluentLogger;
+import com.google.gson.Gson;
+import com.google.maps.GeoApiContext;
+import com.google.maps.PlaceDetailsRequest;
+import com.google.maps.errors.InvalidRequestException;
+import com.google.maps.model.PlaceDetails;
+import com.google.tripmeout.frontend.PlaceVisitModel;
+import com.google.tripmeout.frontend.storage.PlaceVisitStorage;
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.util.List;
+import java.util.Optional;
+import java.util.regex.Pattern;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+public class PlaceParentServlet extends HttpServlet {
+  private static final FluentLogger logger = FluentLogger.forEnclosingClass();
+  private final PlaceVisitStorage placeStorage;
+  private final Gson gson;
+
+  private static final Pattern TRIP_NAME_PATTERN = Pattern.compile(".*/trips/([^/]+)/placeVisits");
+
+  private final GeoApiContext CONTEXT =
+      new GeoApiContext.Builder().apiKey("AIzaSyBbXCXC2uWv3baNmirLtqUYbFsFwCXqLV8").build();
+
+  public PlaceParentServlet(PlaceVisitStorage placeStorage, Gson gson) {
+    this.placeStorage = placeStorage;
+    this.gson = gson;
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    try {
+      Optional<PlaceVisitModel> place =
+          ServletUtil.extractFromRequestBody(request, gson, PlaceVisitModel.class);
+
+      String tripId = ServletUtil.matchUriOrThrowError(request, TRIP_NAME_PATTERN).group(1);
+      logger.atInfo().log("finished matching");
+
+      if (!place.isPresent()) {
+        response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+        return;
+      }
+
+      PlaceDetailsRequest placeRequest =
+          new PlaceDetailsRequest(CONTEXT)
+              .placeId(place.get().placeId())
+              .fields(PlaceDetailsRequest.FieldMask.GEOMETRY_LOCATION,
+                  PlaceDetailsRequest.FieldMask.NAME);
+
+      PlaceDetails details;
+      details = placeRequest.await();
+
+      PlaceVisitModel newPlace = PlaceVisitModel.builder()
+                                     .setTripId(tripId)
+                                     .setPlaceId(place.get().placeId())
+                                     .setUserMark(PlaceVisitModel.UserMark.YES)
+                                     .setName(details.name)
+                                     .setLatitude(details.geometry.location.lat)
+                                     .setLongitude(details.geometry.location.lng)
+                                     .build();
+
+      placeStorage.updateUserMarkOrAddPlaceVisit(newPlace, PlaceVisitModel.UserMark.YES);
+      PlaceVisitModel newOrUpdatedPlace =
+          placeStorage.getPlaceVisit(tripId, place.get().placeId()).get();
```

Comment:
The alarm bells should be ringing whenever you call `get()` on an optional without first checking it. In this case, it _may_ actually fail during a race with a separate request that deletes the given trip/place.

This is actually why I was asking about how you intended to use the update method when designing the storage API. Looks like we don't actually care about whether or not an existing place was updated, but just want the end result itself.

`updateUserMarkOrAddPlaceVisit` should probably be updated to return the updated (or newly inserted) place visit object. That _guarantees_ that this call returns a valid result (modulo any legitimate storage access errors).

If you don't want to fix that in this PR, then please add a TODO and a new issue to make that refactor (which also includes updating all callsites).